timeseries_storage/blueflood/blueflood.go

```diff
@@ -450,6 +450,44 @@ var samplerMap map[api.SampleMethod]sampler = map[api.SampleMethod]sampler{
 	},
 }
 
+// Blueflood will use the finest-grained resolution which doesn't exceed the slot limit.
+// Thus, if you request too many points, it will automatically downsample the data.
+func (b *Blueflood) AdjustTimerange(requested api.Timerange, slotLimit int) api.Timerange {
+	// In some cases, coarser-resolution data may have a shorter TTL.
+	// To accomodate these cases, it must be verified that the requested timerange will
+	// actually be present for the chosen resolution.
+	// TODO: figure out how to make this work with moving averages and timeshifts
```

Comment:
It's possibly a bit confusing, but `AdjustTimerange` is performed at the *beginning* of each `select` before evaluating the body, *not* before each fetch.

The reason for this is (1) none of the functions can tolerate working with data that has differing resolutions (aggregation and joins are much more complicated if the resolution is different in different parts of the data) and (2) no part of the code but the timeseries storage knows what resolutions are valid for which time periods.

Basically, when the user performs a query, the first thing that happens is that the resolution will be changed if the `TimeseriesStorageAPI` knows that the user-selected resolution won't work.